pkg/storegateway/bucket_index_reader.go

```diff
@@ -572,13 +569,23 @@ func newBucketIndexLoadedSeries() *bucketIndexLoadedSeries {
 	}
 }
 
-// loadSeriesForTime populates the given symbolized labels for the series identified by the reference if at least one chunk is within
+// addSeries stores a series raw data in the data structure. A stored series can be loaded via unsafeLoadSeriesForTime().
+// This function is concurrency safe.
+func (l *bucketIndexLoadedSeries) addSeries(ref storage.SeriesRef, data []byte) {
+	l.seriesMx.Lock()
+	l.series[ref] = data
+	l.seriesMx.Unlock()
+}
+
+// unsafeLoadSeriesForTime populates the given symbolized labels for the series identified by the reference if at least one chunk is within
 // time selection.
-// loadSeriesForTime also populates chunk metas slices if skipChunks if set to false. Chunks are also limited by the given time selection.
-// loadSeriesForTime returns false, when there are no series data for given time range.
+// unsafeLoadSeriesForTime also populates chunk metas slices if skipChunks if set to false. Chunks are also limited by the given time selection.
+// unsafeLoadSeriesForTime returns false, when there are no series data for given time range.
 //
 // Error is returned on decoding error or if the reference does not resolve to a known series.
-func (l *bucketIndexLoadedSeries) loadSeriesForTime(ref storage.SeriesRef, lset *[]symbolizedLabel, chks *[]chunks.Meta, skipChunks bool, mint, maxt int64, stats *queryStats) (ok bool, err error) {
+//
+// This function is NOT concurrency safe.
```

Comment:
wouldn't `not concurrency safe` mean that it can't be called multiple times concurrently? afaict the problem is not that this function can't be called multiple times concurrently, I think actually that would be fine because it only reads, but it can't be called concurrently with `addSeries()`, so I would make the comment a bit more explicit about that.

Reply:
Good point. Changed the comment in https://github.com/grafana/mimir/pull/3500/commits/a2b06154d76d8c20dc9d3522fe96d365fc642461.